### ISSUE TITLE
[RFE] compile modified files first

### ISSUE THREAD

**Issue by Hi-Angel** [ninja-build/ninja#1570]
**Problem**: if a project has 300 cpp files and a single header that's included everywhere, and then someone changed the header and a single cpp file, most likely one wants to know whether changes in these 2 files caused any build errors. However, after executing `ninja -C build`, in the worst case, the changed cpp file may get built after the 299 others, and fail the build.

Ultimately, when rebuilding a project, one is interested *(well, besides of course having the binaries)* in "did my changes cause any problem?" question.

**Solution**: before building, sort files to build by "last changed" timestamp, and compile changed ones first.

**Comment by Hi-Angel** [comment]
I'm also interested in whether PRs to implement this are welcome. I don't promise to make one, but there are situations when I'm bored and I could take a look.

I'm asking this explicitly, because I see this project has a lot of PRs that don't get acted upon by anybody.

**Comment by jhasse** [comment]
Changing the build order has been discussed a lot before, see #376 for example. Someone brought up the exact same scenario you're describing: https://github.com/ninja-build/ninja/issues/376#issuecomment-88149246

PR would be welcome :)

**Comment by jonesmz** [comment]
This would be an excellent improvement.

**Comment by ilyapopov** [comment]
I would consider a more general problem of build order strategy. There are a few different things you could optimize for. So, I came up with:

1. Optimize for determinism, repeatability and debugging of the build system itself:
   - sort targets in a fixed deterministic order, for example alphabetically by path

2. Optimize for edit-compile cycle
   a. Failed first. Build first target that failed last time.
   b. Last modified first.

3. Optimize for total compile time.
   a. Deepest first. Build first targets that are deepest in the dependency tree
   b. Longest target first. From the log file we know how much time it took for a target to build last time. We can use this information. 
   c. Longest path first. The same, but we use not the time it took to build the target itself, but the whole path to the tree top.

Obviously, all strategies except 1 can be combined in various ways.

I am not arguing to implement all of these, just things to consider.

**Comment by Hi-Angel** [comment]
> 3. Optimize for total compile time.

@ilyapopov I've read at the linked discussion about building most-time-consuming target first, but didn't quite get it: how can this improve build time? As I understand, each target being built independently. Even when they share sources, the sources would be built multiple times, per each target.

So I don't see how reordering of targets could influence building time.

**Comment by Hi-Angel** [comment]
> Even when they share sources, the sources would be built multiple times, per each target.

Nvm, about this one, it's specific to meson I guess. The point still stands though.

**Comment by ilyapopov** [comment]
The idea is to avoid situation where in the end you have one long task running while all other CPUs and idling.

Imagine you have 8 tasks taking 1s each and a task taking 3s. You execute on 4-core CPU. If you schedule short tasks first, the whole process would take 5 sec. If you schedule long task first, the whole process will take only 3 sec, because short task will be executed in parallel with the long one.

**Comment by ilyapopov** [comment]
See this comment https://github.com/ninja-build/ninja/issues/376#issuecomment-330909603 for the illustration of such a scenario.

**Comment by Hi-Angel** [comment]
So, I spent some time getting acknowledged with the code. The code isn't easy, a lot of stuff going on that I can't understand. I figured though that `struct Plan` stores targets to build, and I'd need to sort them by mtime in the plan.

The first problem I see: `Plan` stores the plan as a graph, and traversing the graph *(e.g. to sort)* gonna take a lot of CPU time.

What is the reason `Plan` stores tasks to do as a graph instead of just using a vector?

**Comment by Hi-Angel** [comment]
Aha, my experiment shows that I probably don't need to traverse the tree.

The `Plan` has `want_` variable that is a map with `Edge*`s. For each such Edge, its input nodes are files to compile *(e.g. includes, .cpp, c, etc)*, and output nodes has usually just one element, which is the target of compilation of these files *(e.g. an object file, an executable, a library…)*.

Presumably, all I need to do is to sort the map. Perhaps I could try providing to it a custom comparator that compares `mtime`. I don't know how good of an idea it is, especially given that Edges do not have mtime, so I also need to code means for providing it.

**Comment by Hi-Angel** [comment]
> Perhaps I could try providing to it a custom comparator that compares mtime. I don't know how good of an idea it is, especially given that Edges do not have mtime, so I also need to code means for providing it.

Nah, apparently this not gonna work, because from what I've just read if the comparator is used not just for sorting, but for determining uniqueness.

Do you think maybe worth to migrate `want_` from an `std::map` to a plain vector? I don't see any other solution.

**Comment by jhasse** [comment]
> Nah, apparently this not gonna work, because from what I've just read the comparator is used not just for sorting, but rather for determining uniqueness.

Compare by mtime first and then by the pointer?

**Comment by Hi-Angel** [comment]
Thanks!

So, FTR, [I've implemented sorting the map by mtime](https://github.com/Hi-Angel/ninja/commit/eed2664ae42eb8926336d8a1ff0e96ed43f92178), and I checked that the sorting works. However it does not have any influence on the build order whatsoever.

Ideas are welcome. I'm wondering, if it has something to do with ["work" being taken from `ready_` field](https://github.com/ninja-build/ninja/blob/20b30dac6698d119e7797b34d6ed2c4ed8f48417/src/build.cc#L378) *(although what work could be taken from edges that are ready already, I don't know…)*.

**Comment by Hi-Angel** [comment]
> Ideas are welcome. I'm wondering, if it has something to do with ["work" being taken from `ready_` field](https://github.com/ninja-build/ninja/blob/20b30dac6698d119e7797b34d6ed2c4ed8f48417/src/build.cc#L378) _(although what work could be taken from edges that are ready already, I don't know…)_.

Though [sorting `ready_` instead](https://github.com/Hi-Angel/ninja/commit/aec05a76919c4cf851b55ab29c7d7670cf8c0bdc) indeed does something to build order, however most recently modified files not necessarily go first.

**Comment by Hi-Angel** [comment]
> > Ideas are welcome. I'm wondering, if it has something to do with ["work" being taken from `ready_` field](https://github.com/ninja-build/ninja/blob/20b30dac6698d119e7797b34d6ed2c4ed8f48417/src/build.cc#L378) _(although what work could be taken from edges that are ready already, I don't know…)_.
> 
> Though [sorting `ready_` instead](https://github.com/Hi-Angel/ninja/commit/aec05a76919c4cf851b55ab29c7d7670cf8c0bdc) indeed does something to build order, however most recently modified files not necessarily go first.

Could it be because ninja runs multiple targets simultaneously, and those "other" files, that I see sometimes first, just happen to be one of them, together with the "last modified" target?

This nuance aside, the change to `ready_` field definitely pops last modified files up in the build order.

**Comment by Hi-Angel** [comment]
> Could it be because ninja runs multiple targets simultaneously, and those "other" files, that I see sometimes first, just happen to be one of them, together with the "last modified" target?
>
> This nuance aside, the change to `ready_` field definitely pops last modified files up in the build order.

lol I should've tried setting `-j 1` to only run one target at a time.

Yeah, sorting the `ready_` field definitely works. [PR sent](https://github.com/ninja-build/ninja/pull/1577).
